### PR TITLE
Combined all fixes

### PR DESCRIPTION
**Pull request by chryswoods**
## If this is to fix a bug...

This pull request fixes issue #26, plus the slow AmberPrm issue (fix_amberprm_parallel), plus compile issues on MacOS with older boost (fix_mcs_unary_function) plus the version check on compile (feature_version_check) (which I would call a fix rathe than a feature ;-)).

I have also taken the opportunity to update the GitHub actions workflows and split them into three:

1. pr.yaml - for pull requests only. Not manually runnable. Only builds the pull request for the latest python on Windows, Mac and Linux, and only older Pythons on Linux. This should catch everything except for weird errors on older Pythons on non-Linux systems. It saves 4 out of 9 builds per pull request.

2. devel.yaml - for pushes to devel or manual runs only. Only builds the devel branch for the latest python on Windows, Mac and Linux, and only older Pythons on Linux. This will upload the packages automatically to anaconda cloud. This should catch all but the weird errors, saves 4 out of 9 builds per pull request, and will save space on anaconda cloud as now only the latest Python version will be supported on devel for Mac and Windows.

3. main.yaml - only for manual runs, for which you have to say whether or not you want to upload packages to anaconda cloud (yes / no). This runs everything on the main branch. We can use this to test that fixes / releases do build and pass all tests on all supported platforms and versions, and can choose whether or not to upload the packages based on whether or not we are running a release. Eventually I would like to split this into two, with the run creating packages that are held in GitHub Actions as artefacts, and then the second doing the actual upload if we choose to and the artefacts have been created correctly.

* I confirm that I have merged the latest version of `devel` into this branch before issuing this pull request (e.g. by running `git pull origin devel`): [y]
* I confirm that I have permission to release this code under the GPL3 license: [y]

## Suggested reviewers:
@lohedges

## Any additional context of information?

I have tested the loading and saving of the SLipids molecules, but can't add any nice tests until I have added in the code to support the forcefield terms in the internal forcefield. I think the code to support reading/writing is a "fix" and can be in 2023.2.1, while the code to add the internal forcefield terms and make sure that these all work correctly (including in openmm) is a "feature" for 2023.3.0. I don't want to block the fix from going in, so it is good to merge. I will do the feature on a proper feature branch.

### CODE REVIEW

**Comment by lohedges** [comment]
Thanks for this. I agree that my edit is certainly a fix not a feature. Not sure what I was thinking there.

With the changes to the CI I have a question about tags and how we make things persistent. At present we are tagging on `devel` and merging releases across to `main`, so the branches are identical at that point and tags point to the same commit. With the new approach, I'm assuming that fixes get merged into `devel` and `main`, but only `main` gets the patch release tag and the new patch build? If we are going to build patch releases for `main` and `devel` then we would probably need to start tagging `main` and `devel` differently for patch releases, e.g. `2023.2.1` on `main` and `2023.2.1-dev` on `devel`. However, at the point of a full release we will now need to do the work to fix any conflicts when merging `devel` into `main`. We can't simply blat `main` and replace with `devel` since we would lose the commits and tags for the `main` patch releases, which would remove the releases on GitHub. (`conda` packages would be fine, though.)

I'll mirror the same approach for BioSImSpace once we are happy with things.

File: .github/workflows/pr.yaml [hunk 1]
```diff
@@ -0,0 +1,76 @@
+name: Pull-Request
+
+# Note that push and pull-request builds are automatically
+# now skipped by GitHub if
+# [skip ci], [ci skip], [no ci], [skip actions], or [actions skip]
+# are in the commit message. We don't need to check for this ourselves.
+
+on:
+  pull_request:
+    branches: [devel]
```
Comment (lohedges) [inline .github/workflows/pr.yaml:10]: Do we not want to check builds for fix PRs against `main` too? I doubt `main` and `devel` would be so different that the application of the fix would be drastically different between the two branches, but I could imagine cases where something might be subtly different if an new feature has already been applied on `devel` before the fix.

Comment (chryswoods) [reply .github/workflows/pr.yaml:10]: Yes - you are right that I missed this ;-)

**Comment by chryswoods** [comment]
Yes - I haven't quite worked out how tags are going to work. This is the first stage in experimenting to find something thing that is workable. I like the idea of having a 'dev' tag for dev branches, especially as it will then be clearer that dev code is being used. I think main will drift away from devel as we make patch releases, and then we will need to be careful to pull from devel to main and fix any issues / conflicts when we make a main major release.

**Comment by lohedges** [comment]
Great, that makes sense. Was checking that I wasn't misunderstanding anything :-) I don't think conflicts will be too bad. Since we are applying fixes in both places then we should be able to assume the `devel` is the single source of truth at the point of the release, so can simply replace conflicted on `main` files with those `devel` at the point of release. Perhaps missing some edge cases here :thinking: As you say, it will probably take a bit of experimenting to find a process that works.

**Pull request merged**